### PR TITLE
Just in time OBJ/IR Caching

### PR DESCRIPTION
**Pull request by skmp**
#### Overview

Caches IR and OBJ as it gets compiled, in a multithreaded and multi process safe way.

This started as a cleanup of the loading interface, and as such has a lot of cruft, then I cherry-picked stuff in from #1548 

Reading from the caches takes a shared_lock, and that's it.

Writing to the caches takes a unique_lock, and also fnctl advisory lock when the index or the data files are resized.

Index files are re-mapped as they grow, they grow in 64kb chunks.

Data files are mapped in 'chunks' of up to 16Megs.

Settings have been cleaned up. `AOTIRLoad` and `AOTIRCapture` have been merged to `IRCache`.  `AOTIRGenerate` has been completely removed, and a separate executable, `FEXAOTGen` is used for that.

To enable caching, use
```
"IRCache": "auto"/"disabled"/"read"/"write"/"readwrite",
"ObjCache": "auto"/"disabled"/"read"/"write"/"readwrite"
```

This is ready for review, and I propose we merge it as an experimental feature for 2209, with follow ups, further testing and stability fixes planned for 2210, 2211 and 2212

#### Todo
- [x] Rework in a generic index file
- [x] Cleanup & Implement configuration
- [x] Cleanup object cache / debug changes
- [x] See what broke tests
- [x] Update Summary
- [x] Investigate glmark2 rare crash
- [x] gimp also crashes

#### Follow Up
- [x] Fix `#if FIXME`s (mostly around the unused visual debugger, #1960)
- [x] Make sure thunk relocations work (#1842)
- [x] Investigating adding some form of balancing in the BSTs? (#1936)
- [x] Investigate adding a sorted array mode if the BSTs are not dirty (#1936)
- [x] Further optimise INDEX and DATA advisory locks (#1962)

### CODE REVIEW

**Comment by skmp** [comment]
Among other thigs, this also hashes the actual block ranges, not (start, end). It also limits the blocks to a single vma mapping.

**Comment by skmp** [comment]
Also, it inserts ranges for cached blocks for SMC invalidation, though only after hashing them.

**Comment by skmp** [comment]
(Refactor broke this branch, will fix toms)

**Comment by skmp** [comment]
Steam is stable again :)

**Comment by skmp** [comment]
(now with OBJ support for x86 and arm64)

**Comment by skmp** [comment]
This has been rebased on top of main, and most bugs have been fixed. I plan to spent tomorrow on cleanups as well.

Would be good to get some feedback as-is @Sonicadvance1 @neobrain ~

**Comment by Sonicadvance1** [comment]
Would be nice if this passed CI before reviewing thoroughly.

**Comment by skmp** [comment]
> Would be nice if this passed CI before reviewing thoroughly.

The CI fails are unrelated to the main logic, it's just not writed initialization for the TestHarness. I'll be pushing a fix in a bit

**Comment by skmp** [comment]
This has been cleaned up, with most dead/stale code removed.

Still a few items to do
- [ ] Fully postfix cache folders with compilation modal options

File: External/FEXCore/Source/Interface/Context/Context.cpp [hunk 1]
```diff
@@ -171,38 +174,32 @@ namespace FEXCore::Context {
...
+  void UnloadNamedRegion(FEXCore::Context::Context *CTX, Core::NamedRegion *Entry) {
+    return CTX->UnloadNamedRegion(Entry);
   }
 
   CustomIRResult AddCustomIREntrypoint(FEXCore::Context::Context *CTX, uintptr_t Entrypoint, std::function<void(uintptr_t Entrypoint, FEXCore::IR::IREmitter *)> Handler, void *Creator, void *Data) {
     return CTX->AddCustomIREntrypoint(Entrypoint, Handler, Creator, Data);
   }
 
 namespace Debug {
+  #if FIXME
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Context/Context.cpp:202]: Fix this

**Comment by skmp** [comment]
- [ ] Fix `#if FIXME`s (mostly around the unused visual debugger)
- [ ] Make AOTGenerate a fexloader-only option

**Comment by skmp** [comment]
- [ ] Fix FEXUpdateIRCache.sh, add  FEXUpdateObjCache.sh, cleanup scripts?
- [ ] Maybe merge them to a python script?

**Comment by skmp** [comment]
- [ ] Compact relocations to actual size
- [ ] Make sure thunk relocations work

Follow Up:
- [ ] Generate relocations in relocation pools, import cached files directly to code cache

**Comment by Sonicadvance1** [comment]
All tasks completed on this? Merge conflict is still here.

**Comment by skmp** [comment]
(This has been squashed and rebased to main)

**Comment by skmp** [comment]
After much, much debugging and fixing several other bugs, it looks like this is a race condition that goes away when code isn't compiled 'too fast'.

So far I've verified that it is not
- Bad code stored in cache
- Bad code read from cache
- Cache index/data corruption

**Comment by skmp** [comment]
@Sonicadvance1 thoughts on disabling/ignoring the `Visual Debugger` for now? It is largely broken at this point, all of the APIs i've disabled here were already broken before.

Otherwise, I'll do another bug hunting spree for this tonight, and if the bug is not found I think it's best to merge and keep experimental til 2210.

**Comment by skmp** [comment]
Bug fixed, ranges were not serialized correctly.

Performance drops a little with that, though we can work around it in different ways in the future, such as whole page caching.

Now investigating the interpreter failures.

**Comment by skmp** [comment]
With the last round of fixes this passes all tests for me locally.

Assuming it passes CI, I'll take another look tomorrow to cleanup things and get everything ready for merge.

**Comment by skmp** [comment]
Looks like all the fixes have managed to round up smc issues outlined in #1754 at last.

Based on my testing, parallels/m1 is ~ 60% likely to fail on that test, with stale code running for ever.

Oddly enough, it doesn't repro in orion

**Comment by skmp** [comment]
The smc issues don't repro with gdb attached, so likely yet another race somewhere between multi threaded invalidation and translation.

Also, the smc test crash with objc enabled, possibly with irc as well. While not a blocker, I'll look into that as well.

**Comment by skmp** [comment]
Hmm, looks like that resolved the SMC issue, though we got another spurious failure in `pthread_cancel` in the 8.4 runner. I've seen this one before, though I haven't investigated it. Logged as follow up for https://github.com/FEX-Emu/FEX/issues/1754#issuecomment-1232677973 and will re-run CI here.

**Comment by skmp** [comment]
> Also, the smc test crash with objc enabled, possibly with irc as well. While not a blocker for now, possibly follow up?

Generated #1958

File: External/FEXCore/Source/Interface/Core/Dispatcher/X86Dispatcher.cpp [hunk 2]
```diff
@@ -169,7 +169,7 @@ X86Dispatcher::X86Dispatcher(FEXCore::Context::Context *ctx, const DispatcherCon
     ret();
   }
 
-  constexpr bool SignalSafeCompile = true;
+  constexpr bool SignalSafeCompile = false;
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/Dispatcher/X86Dispatcher.cpp:172]: This needs to be turned back on

File: External/FEXCore/Source/Interface/Core/Interpreter/MiscOps.cpp [hunk 3]
```diff
@@ -77,7 +79,7 @@ DEF_OP(GetRoundingMode) {
     mrs %[Tmp], FPCR;
   )"
   : [Tmp] "=r" (Tmp));
-  // Extract the rounding
+  // `tract the rounding
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/Interpreter/MiscOps.cpp:82]: This needs to be reverted

File: External/FEXCore/Source/Interface/Core/JIT/x86_64/MiscOps.cpp [hunk 4]
```diff
@@ -46,10 +46,17 @@ DEF_OP(Fence) {
 DEF_OP(Break) {
   auto Op = IROp->C<IR::IROp_Break>();
 
+  lea(rax, ptr[rip]);
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/JIT/x86_64/MiscOps.cpp:49]: These needs to hide behind a debug option

File: External/FEXCore/Source/Interface/Core/LookupCache.cpp [hunk 5]
```diff
@@ -30,6 +30,7 @@ LookupCache::LookupCache(FEXCore::Context::Context *CTX)
   // We need one pointer per page of virtual memory
   // At 64GB of virtual memory this will allocate 128MB of virtual memory space
   PagePointer = reinterpret_cast<uintptr_t>(FEXCore::Allocator::mmap(nullptr, ctx->Config.VirtualMemSize / 4096 * 8, PROT_READ | PROT_WRITE, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0));
+  LOGMAN_THROW_A_FMT(PagePointer != -1ULL, "Failed to allocate page pointer");
```
Comment (skmp) [inline External/FEXCore/Source/Interface/Core/LookupCache.cpp:33]: This needs to be `UINTPTR_MAX` or `INVALID_PTR`

File: External/FEXCore/Source/Interface/Context/Context.h [hunk 6]
```diff
@@ -188,30 +191,38 @@ namespace FEXCore::Context {
     void RemoveCustomIREntrypoint(uintptr_t Entrypoint);
 
     // Debugger interface
+#if FIXME
```
[the remark this answers is not in the document]
Comment (skmp) [reply External/FEXCore/Source/Interface/Context/Context.h:212]: Generated https://github.com/FEX-Emu/FEX/issues/1960

File: External/FEXCore/Source/Interface/Core/Interpreter/MiscOps.cpp (hunk 3 above)
Comment (skmp) [reply External/FEXCore/Source/Interface/Core/Interpreter/MiscOps.cpp:82]: Done

File: External/FEXCore/Source/Interface/Core/Dispatcher/X86Dispatcher.cpp (hunk 2 above)
Comment (skmp) [reply External/FEXCore/Source/Interface/Core/Dispatcher/X86Dispatcher.cpp:172]: Done

File: External/FEXCore/Source/Interface/Core/LookupCache.cpp (hunk 5 above)
Comment (skmp) [reply External/FEXCore/Source/Interface/Core/LookupCache.cpp:33]: Done

File: External/FEXCore/Source/Interface/Core/JIT/x86_64/MiscOps.cpp (hunk 4 above)
Comment (skmp) [reply External/FEXCore/Source/Interface/Core/JIT/x86_64/MiscOps.cpp]: Done

**Comment by skmp** [comment]
Some performance numbers

No Cache
```
skmp@ornio:~/projects/FEX/build$ time Bin/FEXLoader /bin/ls > /dev/null

real	0m0,277s
user	0m0,249s
sys	0m0,025s

```
OBJCache
```
skmp@ornio:~/projects/FEX/build$ time Bin/FEXLoader /bin/ls > /dev/null
[Info] Warning: OBJ/IR Caches are experimental, and might lead to crashes.

real	0m0,029s
user	0m0,012s
sys	0m0,015s
```
Native
```
skmp@ornio:~/projects/FEX/build$ time /bin/ls > /dev/null

real	0m0,007s
user	0m0,000s
sys	0m0,007s
```

**Comment by skmp** [comment]
(Closing this as there is an ongoing powergrab by @Sonicadvance1, I will migrate my work to https://github.com/skmp/fex-emu-ng.git)

**Comment by Brickmaster1** [comment]
Why was this not accepted? I am very confused, caching seems like an incredible thing for this project.

**Comment by Sonicadvance1** [comment]
> Why was this not accepted? I am very confused, caching seems like an incredible thing for this project.

It's being worked on, don't worry about it.

**Comment by Brickmaster1** [comment]
Hm, okay. I just saw this repository https://gitlab.com/hex-emu/hex-emu 
and I thought that the developer initially working on the AOT and big LLVM one-time codegen optimizations had splintered off or something. There is no evidence of arguments here as far as I can see. Not sure. I just see the second project had this merged in and was wondering if you had an explanation for why that didn't happen here.

**Pull request closed without merging**